src/slidedown/Slidedown.ts

```diff
@@ -0,0 +1,225 @@
+        icon,
+        positiveButtonText,
+        negativeButtonText
+      });
+
+      // Insert the container
+      addDomElement('body', 'beforeend', `<div id="${SlidedownCssIds.container}"` +
+        `class="${SlidedownCssClasses.container} ${SlidedownCssClasses.reset}"></div>`);
+      // Insert the dialog
+      addDomElement(this.container, 'beforeend',
+          `<div id="${SlidedownCssIds.dialog}" class="${SlidedownCssClasses.dialog}">${slidedownHtml}</div>`);
+
+      // Animate it in depending on environment
+      addCssClass(this.container, bowser.mobile ? 'slide-up' : 'slide-down');
+
+      // Add click event handlers
+      this.allowButton.addEventListener('click', this.onSlidedownAllowed.bind(this));
+      this.cancelButton.addEventListener('click', this.onSlidedownCanceled.bind(this));
+      Event.trigger(Slidedown.EVENTS.SHOWN);
+    }
+  }
+
+  async onSlidedownAllowed(_: any) {
+    await Event.trigger(Slidedown.EVENTS.ALLOW_CLICK);
+  }
+
+  onSlidedownCanceled(_: any) {
+    Event.trigger(Slidedown.EVENTS.CANCEL_CLICK);
+    this.close();
+  }
+
+  close() {
+    addCssClass(this.container, 'close-slidedown');
+    once(this.dialog, 'animationend', (event: any, destroyListenerFn: () => void) => {
+      if (event.target === this.dialog &&
+          (event.animationName === 'slideDownExit' || event.animationName === 'slideUpExit')) {
+          // Uninstall the event listener for animationend
+          removeDomElement(`#${SlidedownCssIds.container}`);
+          destroyListenerFn();
+          Event.trigger(Slidedown.EVENTS.CLOSED);
+      }
+    }, true);
+  }
+
+  /**
+   * only used with Category Slidedown
+   */
+  toggleSaveState() {
+    if (!this.categoryOptions) {
+      Log.debug("Slidedown private category options are not defined");
+      return;
+    }
+
+    if (!this.isInSaveState) {
+      // note: savingButtonText is hardcoded in constructor. TODO: pull from config & set defaults for future release
+      this.allowButton.innerHTML = this.getIndicatorHolderHtmlWithText(this.categoryOptions!.savingButtonText!);
+      addDomElement(this.buttonIndicatorHolder, 'beforeend',
+        getLoadingIndicatorWithColor(COLORS.whiteLoadingIndicator));
+      (<HTMLButtonElement>this.allowButton).disabled = true;
+      addCssClass(this.allowButton, 'disabled');
+      addCssClass(this.allowButton, SlidedownCssClasses.savingStateButton);
+    } else {
+      // positiveUpdateButton should be defined as written in MainHelper.getSlidedownPermissionMessageOptions
+      this.allowButton.innerHTML = this.categoryOptions!.positiveUpdateButton!;
+      removeDomElement(`#${SlidedownCssClasses.buttonIndicatorHolder}`);
+      (<HTMLButtonElement>this.allowButton).disabled = false;
+      removeCssClass(this.allowButton, 'disabled');
+      removeCssClass(this.allowButton, SlidedownCssClasses.savingStateButton);
+    }
+    this.isInSaveState = !this.isInSaveState;
+  }
+
+  toggleFailureState() {
```

Comment:
Same applies to this as my `toggleSaveState` comment above

Reply:
Done.